### ISSUE TITLE
Define order of hooks

### ISSUE THREAD

**Issue by Globegitter** [balderdashy/sails#2517]
I have just written a [traceur](https://github.com/google/traceur-compiler) hook which I want to use to get ES6/7 access throughout my sails app.
It mostly seems to work, but not for my services. I suppose that is because the services get loaded before the traceur-hook gets loaded.

Is there any way to define a hook order in an app? Or maybe set in the hook that it should get loaded early, or before a certain default-hook?

This is using sails 0.11 of course.

Edit: Ahh, if I write ES6/7 code directly in the controllers it doesn't work either (since they also get loaded before traceur I assume). But code that I ES6/7 code that I require within the controller works fine (since traceur is overriding Nodejs' require with it's own)

Edit2: If I manually add the hook as one of the first ones I seem to get a step further, but now running into a seemingly unrelated issue.

Edit3: Yep, got it all working with adding it into the logger-hook (to test quickly) and applying this fix https://github.com/balderdashy/waterline/pull/776
So the original issue still stands and we could get immediate ES6/7 support in sails.

Oh and here is the repository itself: https://github.com/artificialio/sails-hook-traceur

**Comment by sgress454** [comment]
Hey @Globegitter, sorry about the lack of response on this, the hook looks awesome.  So was there anything you had to do to get this to work besides the Waterline fix?

Re: hook load order, there have been some suggestions in this area but nothing's been implemented yet.  It's a bit of a sticky wicket, that one...

**Comment by Globegitter** [comment]
@sgress454 No worries, I do understand. Well if you want to write ES6/7 straight in the Controllers/Services/etc. you do need to be able to define the hook order.

If on the other hand you use the Controllers, etc. to require an ES6/7 file (and then just call the functions as usual) it does work if I remember correctly. Thanks for having the fix merged btw. So do you think this functionality is likely to be in for 0.11, or do you think it would push it too far back?

**Comment by sgress454** [comment]
Changing hook load order through configuration will not be available in v0.11.x.  I looked at `sails-hook-traceur` though, and seeing as it doesn't do anything asynchronous in its `.initialize()`, you might be able to get away with just moving all that logic into `.configure()`.  The `.configure()` methods all run *before* any `.initialize()` methods, which means that services, controllers, etc wouldn't be loaded until after traceur was activated.  Worth a shot as a workaround.

**Comment by Globegitter** [comment]
Aahh, managed to get it to work that way @sgress454 - thank you!

**Comment by davepreston** [comment]
Any progress on hook load order through configuration?

In my case I have two hooks. One that determines who the user is (authentication) and the second that handles authorization. Since this is per user it needs to be ordered in the "routes" phase of the hook. At this point the only workaround I can come up with is to combine the logic from my two hooks into one hook and manage the order inside my "mother" hook.

**Comment by daedalus28** [comment]
@davepreston You can use the hook loaded events to tell when another hook is loaded - in your case, have the second hook wait for the first in its initialize function. Here's a link to the relevant section of the documentation.
http://sailsjs.org/documentation/concepts/extending-sails/hooks/hook-specification/initialize#?hook-events-and-dependencies

However, this doesn't seem to cover the case of requiring a hook to load before another one - only after. Meaning if you don't control the code of the hook that loads afterwards (such the core services and controllers hooks), there is no clear way to make sure you module loads first.

**Comment by sailsbot** [comment]
Thanks for posting, @Globegitter.  I'm a repo bot-- nice to meet you!

It has been 60 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by danywheeler** [comment]
well, is this has been resolved? I think that sometimes we do need to specify the hook load order for some rewrites to work!

**Comment by emahuni** [comment]
I think that sails bot needs to be removed. It is closing open issues, thereby leaving unresolved issues such as this.

**Comment by emahuni** [comment]
This issue needs to be resolved.

**Comment by emahuni** [comment]
I figured it out: I changed the name from `sails-hook-polymorphic-orm` to `sails-hook-et-polymorphic-orm`. The reason was that this hook was being loaded before the orm when installed from npm. In tests and `api/hooks` etc it is loaded well before `orm`. So changing the name to start with an `e` makes sure it is loaded before `orm` that start with an `o` and not after. This works but is very brittle. If for some reason the implementation behind loading of hooks changes that will break with it.

All this is because of the lack of `sails.before` method or any other way to load hooks in a particular order. the only way is to rename the hook to have preceedence alphabetically.
